### PR TITLE
Issue 5905: Catch Exception and check UnsupportedOperationException in ChunkStorageTests. (#5905)

### LINKED ISSUE #5905: Unit test ChunkStorageTests.testConcat fails when the storage does not support concat
**Describe the bug**
When a custom implementation of LTS doesn't support concat operations:
- The `supportConcat` returns `false`;
- It throws `UnsupportedOperationException` when it call `concat` function.

The unit test `ChunkStorageTests.testConcat` fails in this case.  

Because `UnsupportedOperationException` is wrapped by `ExecutionException`, thrown by `CompletableFuture`. Therefore,`catch (UnsupportedOperationException)` doesn't work.

**To Reproduce**
1. Implement a custom LTS chunk storage.
2. Throw `UnsupportedOperationException` in `doConcat` function directly.
3. Run the unit test `ChunkStorageTests.testConcat`.
4. The test fails

**Screenshots**

**Additional information**
Add any other information about the problem here, including the root cause if known.

### PR DESCRIPTION
**Pull request by xmuyoo**
Issue 5905: Catch Exception and check UnsupportedOperationException in ChunkStorageTests. (#5905)

Signed-off-by: xmuyoo <xmuyoo@163.com>

**Change log description**  
(2-3 concise points about the changes in this PR. When committing this PR, the committer is expected to copy the content of this section to the merge description box)
- Catch `Exception` in "concat" tests in `ChunkStorageTests`, and then check the cause whether is `UnsupportedOperationException` or not.
- Based on `master` branch.

**Purpose of the change**  
- Fixes #5905 

**What the code does**  
- In `testConcat`, `testConcatNotExists` of `ChunkStorageTests`, it catches `Exception` when the storage does not support concat operations. And then it gets the cause wrapped by CompletableFuture and check whether it's `UnsupportedOperationException` or not. 
- In `testReadOnly` of `ChunkStorageTests`, it fixes the incorrect assertion of the caught exception.

**How to verify it**  
- These tests should pass if it's a custom LTS implementation which doesn't support concat operations.

### CODE REVIEW

**Review by andreipaduroiu — COMMENTED** [review]
LGTM. @sachin-j-joshi

**Comment by andreipaduroiu** [comment]
@xmuyoo Looks like your build failed. Please run spotbugs locally (`./gradlew segmentstore:storage:spotbugs`) and fix the build to make it pass.

File: segmentstore/storage/src/test/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageTests.java [hunk 1]
```diff
@@ -769,7 +778,11 @@ public void testReadonly() throws Exception {
             chunkStorage.delete(hWrite).join();
         } catch (Exception e) {
             val ex = Exceptions.unwrap(e);
-            Assert.assertTrue(ex.getCause() instanceof UnsupportedOperationException);
+            if (ex instanceof ChunkStorageException) {
+                Assert.assertTrue(ex.getCause() instanceof UnsupportedOperationException);
```
Comment (sachin-j-joshi) [inline segmentstore/storage/src/test/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageTests.java:782]: Perhaps using assertTrue makes intent more clear.

`Assert.assertTrue(ex instanceof UnsupportedOperationException || (ex instanceof ChunkStorageException && ex.getCause() instanceof UnsupportedOperationException));`

File: segmentstore/storage/src/test/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageTests.java [hunk 2]
```diff
@@ -549,8 +549,13 @@ public void testConcatNotExists() throws Exception {
                             }
                     ),
                     ex -> ex instanceof ChunkNotFoundException);
-        } catch (UnsupportedOperationException e) {
-            // The storage provider may not have native concat.
+        } catch (Exception e) {
```
Comment (sachin-j-joshi) [inline segmentstore/storage/src/test/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageTests.java:552]: Perhaps using assertTrue makes intent more clear.

`Assert.assertTrue(ex instanceof UnsupportedOperationException);`

File: segmentstore/storage/src/test/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageTests.java [hunk 3]
```diff
@@ -714,9 +719,13 @@ public void testConcat() throws Exception {
                 bytesRead += chunkStorage.read(ChunkHandle.readHandle(tragetChunkName), bytesRead, bufferRead.length, bufferRead, bytesRead).get();
             }
             assertArrayEquals(bufferWritten, bufferRead);
-
-        } catch (UnsupportedOperationException e) {
-            // The storage provider may not have native concat.
+        } catch (ExecutionException e) {
```
Comment (sachin-j-joshi) [inline segmentstore/storage/src/test/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageTests.java:722]: same as above

**Comment by sachin-j-joshi** [comment]
@xmuyoo Thanks a lot for fixing this. 
I have left some comments. 
But otherwise it looks good.

**Comment by xmuyoo** [comment]
> @xmuyoo Thanks a lot for fixing this.
> I have left some comments.
> But otherwise it looks good.

Thanks @sachin-j-joshi It has been updated.

**Comment by xmuyoo** [comment]
The failed tests can be run successfully in my local, with commands: 
```
# For Segment Store Unit Tests
./gradlew segmentstore:server:test segmentstore:storage:test segmentstore:storage:impl:test bindings:test --parallel 

# For Integration Tests
./gradlew test:integration:test
```

So, any suggestions to resolve this?

**Review by sachin-j-joshi — APPROVED** [review]
LGTM.
I retriggered the tests.

**Review by andreipaduroiu — APPROVED** [review]

**Pull request merged**